### PR TITLE
Remove undefined symbols from versioning script

### LINKED ISSUE #50414: Linker script `julia.expmap.in` references symbols that may be not defined
When trying to compile julia master on debian sid using this [Make.user](https://github.com/JuliaLang/julia/files/11952606/Make.user.txt),
if the linker is LLVM LLD 16, there's a bit of a warnings in the output:
```
    LINK /builddirs/julia-dev/usr/lib/libjulia-internal-debug.so.1.11.0
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'asprintf' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'localtime_r' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'add_library_mapping' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_IO_stdin_used' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'llvmGetPassPluginInfo' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'environ' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '__progname' failed: symbol not defined
    LINK /builddirs/julia-dev/usr/lib/libjulia-internal-debug.so.1.11
    LINK /builddirs/julia-dev/usr/lib/libjulia-internal-debug.so
    LINK /builddirs/julia-dev/usr/lib/libjulia-codegen-debug.so.1.11.0
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '__stack_chk_guard' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'asprintf' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'arraylist_grow' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'small_arraylist_grow' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'small_typeof' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'rec_backtrace' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'libsupport_init' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'localtime_r' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'memhash' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'memhash32' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'memhash32_seed' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'memhash_seed' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'restore_signals' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'add_library_mapping' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'jlbacktrace' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'jlbacktracet' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_IO_stdin_used' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_Z24jl_coverage_data_pointerN4llvm9StringRefEi' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_Z22jl_coverage_alloc_lineN4llvm9StringRefEi' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_Z22jl_malloc_data_pointerN4llvm9StringRefEi' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'environ' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '__progname' failed: symbol not defined
    LINK /builddirs/julia-dev/usr/lib/libjulia-codegen-debug.so.1.11
    LINK /builddirs/julia-dev/usr/lib/libjulia-codegen-debug.so
    JULIA /builddirs/julia-dev/usr/lib/julia/corecompiler.ji
```
... and if LLVM LLD 17 is used, those warnings are errors for some reason,
which makes building impossible. This is quite unfortunate.

cc @topolarity @vtjnash

### LINKED ISSUE #54533: Broken linkerscript, julia fails to link with modern lld versions
Building v1.10.3 with 
```
$ ld -v
Debian LLD 18.1.6 (compatible with GNU linkers)
```
fails:
```
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '__stack_chk_guard' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'asprintf' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'rec_backtrace' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'localtime_r' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'restore_signals' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'add_library_mapping' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'jlbacktrace' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'jlbacktracet' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '_Z24jl_coverage_data_pointerN4llvm9StringRefEi' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '_Z22jl_coverage_alloc_lineN4llvm9StringRefEi' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '_Z22jl_malloc_data_pointerN4llvm9StringRefEi' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'llvmGetPassPluginInfo' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'environ' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '__progname' failed: symbol not defined
```

### PR DESCRIPTION
**Pull request by Zentrik**
I've added multiple versioning script as they all link with different objects and so it's difficult to export empty versions of undefined symbols as was suggested in https://github.com/JuliaLang/julia/issues/54533#issuecomment-2122813744. I removed symbols that as far as I can tell are never defined when a particular versioning script is used.

Fixes #50414, fixes #54533.

### CODE REVIEW

File: src/julia.expmap.in [hunk 1]
```diff
@@ -13,30 +13,26 @@
...
     _jl_timing_*;
     LLVMExtra*;
     JLJIT*;
-    llvmGetPassPluginInfo;
 
     /* freebsd */
-    environ;
-    __progname;
+    environ*;
+    __progname*;
```
Comment (topolarity) [inline src/julia.expmap.in:35]: Why the new `*` here?

File: src/flisp/flisp.c [hunk 2]
```diff
@@ -2485,6 +2485,10 @@ int fl_load_system_image(fl_context_t *fl_ctx, value_t sys_image_iostream)
     return 0;
 }
 
+#ifndef _OS_WINDOWS_
```
Comment (topolarity) [inline src/flisp/flisp.c:2488]: Should this be guarding on `HAVE_VASPRINTF` instead?

Also, why did this come up as part of this change?

Comment (Zentrik) [reply src/flisp/flisp.c:2488]: I think you mean `HAVE_ASPRINTF`, but we never define that as far as I can tell. Instead `asprintf.c` is included if we're on windows.

This has come up as we only define asprintf on windows, so to ensure its not undefined it needs to be set.

File: src/julia.expmap.in (hunk 1 above)
Comment (Zentrik) [reply src/julia.expmap.in:35]: I can't figure out how to have `environ` and `progname` be defined. The *, means lld won't throw an undefined error.

**Comment by Zentrik** [comment]
Hmm some of the symbols I removed do seem to be important e.g. https://libc-alpha.sourceware.narkive.com/SevIQmU3/io-stdin-used-stripped-by-version-scripts

**Pull request closed without merging**